File: js/src/memory/ocr.ts
/**
 * OCR engines for the screen-memory daemon.
 *
 * Backends
 * --------
 * - `vision`    macOS Vision framework via a tiny Swift helper compiled once
 *               with `swiftc` to `<home>/bin/vision-ocr-v1` (the same binary
 *               the Python SDK builds, so the two share it).  Offline, ~0.3 s.
 * - `winrt`     Windows.Media.Ocr via PowerShell.  Offline.
 * - `tesseract` tesseract.js with a persistent worker.  Cross-platform; on
 *               first use it downloads the English model (~10 MB) and caches
 *               it under `<home>/memory/tessdata` so later runs are offline.
 *
 * `auto` prefers the native engine on macOS / Windows and tesseract.js
 * elsewhere.
 */

import { execFile } from "node:child_process";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { promisify } from "node:util";
import { memoryDir, resolveHomeDir } from "./config.js";

const exec = promisify(execFile);

export interface OcrEngine {
  readonly backend: string;
  recognize(png: Buffer): Promise<string>;
  close(): Promise<void>;
}

export type OcrPreference = "auto" | "native" | "vision" | "winrt" | "tesseract";

// Keep byte-for-byte in sync with python/opendesk/computer/ocr.py so both
// SDKs can reuse one compiled helper.
const VISION_SWIFT_SRC = `
import Vision
import AppKit

let args = CommandLine.arguments
guard args.count > 1 else { exit(2) }
let url = URL(fileURLWithPath: args[1])
guard let img = NSImage(contentsOf: url),
      let cgImg = img.cgImage(forProposedRect: nil, context: nil, hints: nil)
else { exit(0) }

let req = VNRecognizeTextRequest()
req.recognitionLevel = .accurate
req.usesLanguageCorrection = true
let handler = VNImageRequestHandler(cgImage: cgImg, options: [:])
try? handler.perform([req])
let lines = (req.results ?? []).compactMap { $0.topCandidates(1).first?.string }
print(lines.joined(separator: "\\n"))
`;

const VISION_HELPER_VERSION = "1";

function which(bin: string): boolean {
  const dirs = (process.env["PATH"] ?? "").split(path.delimiter);
  return dirs.some((d) => d && fs.existsSync(path.join(d, bin)));
}

export function visionHelperPath(home?: string): string {
  return path.join(resolveHomeDir(home), "bin", `vision-ocr-v${VISION_HELPER_VERSION}`);
}

/** Compile the Vision helper once with swiftc.  Returns its path or null. */
export async function ensureVisionHelper(home?: string): Promise<string | null> {
  if (process.platform !== "darwin") return null;
  const bin = visionHelperPath(home);
  if (fs.existsSync(bin)) return bin;
  if (!which("swiftc")) return null;
  const dir = path.dirname(bin);
  fs.mkdirSync(dir, { recursive: true });
  const src = path.join(dir, "vision-ocr.swift");
  try {
    fs.writeFileSync(src, VISION_SWIFT_SRC);
    await exec("swiftc", ["-O", "-o", bin, src], { timeout: 180_000 });
    fs.chmodSync(bin, 0o700);
    return fs.existsSync(bin) ? bin : null;
  } catch {
    return null;
  } finally {
    fs.rmSync(src, { force: true });
  }
}

async function withTempPng<T>(png: Buffer, fn: (p: string) => Promise<T>): Promise<T> {
  const p = path.join(os.tmpdir(), `opendesk-ocr-${process.pid}-${Date.now()}-${Math.random().toString(36).slice(2)}.png`);
  fs.writeFileSync(p, png, { mode: 0o600 });
  try {
    return await fn(p);
  } finally {
    fs.rmSync(p, { force: true });
  }
}

class VisionEngine implements OcrEngine {
  readonly backend = "macos-vision";
  constructor(private helper: string) {}
  async recognize(png: Buffer): Promise<string> {
    return withTempPng(png, async (p) => {
      const { stdout } = await exec(this.helper, [p], { timeout: 30_000, maxBuffer: 8 * 1024 * 1024 });
      return stdout.trim();
    });
  }
  async close(): Promise<void> { /* nothing to release */ }
}

class WinRtEngine implements OcrEngine {
  readonly backend = "windows-winrt";
  async recognize(png: Buffer): Promise<string> {
    return withTempPng(png, async (p) => {
      const ps = p.replace(/\\/g, "/");
      const script = `
Add-Type -AssemblyName System.Runtime.WindowsRuntime
[Windows.Storage.StorageFile, Windows.Storage, ContentType=WindowsRuntime] | Out-Null
[Windows.Media.Ocr.OcrEngine, Windows.Foundation, ContentType=WindowsRuntime] | Out-Null
[Windows.Graphics.Imaging.BitmapDecoder, Windows.Graphics, ContentType=WindowsRuntime] | Out-Null
$file = [Windows.Storage.StorageFile]::GetFileFromPathAsync('${ps}').AsTask().Result
$stream = $file.OpenAsync([Windows.Storage.FileAccessMode]::Read).AsTask().Result
$decoder = [Windows.Graphics.Imaging.BitmapDecoder]::CreateAsync($stream).AsTask().Result
$bitmap = $decoder.GetSoftwareBitmapAsync().AsTask().Result
$engine = [Windows.Media.Ocr.OcrEngine]::TryCreateFromUserProfileLanguages()
$result = $engine.RecognizeAsync($bitmap).AsTask().Result
$result.Lines | ForEach-Object { $_.Text }
`;
      const { stdout } = await exec("powershell", ["-NonInteractive", "-Command", script], { timeout: 30_000, maxBuffer: 8 * 1024 * 1024 });
      return stdout.trim();
    });
  }
  async close(): Promise<void> { /* nothing */ }
}

class TesseractEngine implements OcrEngine {
  readonly backend = "tesseract.js";
  private worker: { recognize(img: Buffer): Promise<{ data: { text: string } }>; terminate(): Promise<unknown> } | null = null;
  constructor(private cacheDir: string) {}
  private async ensure() {
    if (this.worker) return this.worker;
    const { createWorker } = await import("tesseract.js");
    fs.mkdirSync(this.cacheDir, { recursive: true });
    // tesseract.js v5: createWorker(langs, oem, options)
    this.worker = await (createWorker as unknown as (
      langs: string, oem?: number, opts?: Record<string, unknown>,
    ) => Promise<typeof this.worker>)("eng", 1, { cachePath: this.cacheDir, logger: () => {} });
    return this.worker!;
  }
  async recognize(png: Buffer): Promise<string> {
    const w = await this.ensure();
    const { data } = await w.recognize(png);
    return data.text.trim();
  }
  async close(): Promise<void> {
    if (this.worker) {
      try { await this.worker.terminate(); } catch { /* ignore */ }
      this.worker = null;
    }
  }
}

/** Which backend `createOcrEngine` would pick, without creating it. */
export function availableBackend(pref: OcrPreference = "auto", home?: string): string | null {
  const nativeMac = process.platform === "darwin" && (fs.existsSync(visionHelperPath(home)) || which("swiftc"));
  const nativeWin = process.platform === "win32" && which("powershell.exe");
  const tess = fs.existsSync(path.join(resolveNodeModules(), "tesseract.js"));
  if (pref === "vision") return nativeMac ? "macos-vision" : null;
  if (pref === "winrt") return nativeWin ? "windows-winrt" : null;
  if (pref === "native") return nativeMac ? "macos-vision" : nativeWin ? "windows-winrt" : null;
  if (pref === "tesseract") return tess ? "tesseract.js" : null;
  if (nativeMac) return "macos-vision";
  if (nativeWin) return "windows-winrt";
  return tess ? "tesseract.js" : null;
}

function resolveNodeModules(): string {
  // Walk up from this file to find node_modules (works from src/ and dist/).
  let dir = path.dirname(new URL(import.meta.url).pathname);
  for (let i = 0; i < 6; i++) {
    const nm = path.join(dir, "node_modules");
    if (fs.existsSync(nm)) return nm;
    dir = path.dirname(dir);
  }
  return "node_modules";
}

/** Build the preferred engine, or null when none is usable. */
export async function createOcrEngine(pref: OcrPreference = "auto", home?: string): Promise<OcrEngine | null> {
  const backend = availableBackend(pref, home);
  if (backend === "macos-vision") {
    const helper = await ensureVisionHelper(home);
    if (helper) return new VisionEngine(helper);
    if (pref !== "auto") return null;
  }
  if (backend === "windows-winrt") return new WinRtEngine();
  if (backend === "tesseract.js" || (pref === "auto" && backend !== null)) {
    return new TesseractEngine(path.join(memoryDir(home), "tessdata"));
  }
  return null;
}

File: python/opendesk/computer/ocr.py
"""Screen-region OCR — extract visible text from any screen area.

Backends (tried in order)
--------------------------
1. ``pytesseract`` — best quality, cross-platform; needs Tesseract binary.
2. macOS Vision framework — zero extra deps on macOS 11+, via Swift subprocess.
3. Windows WinRT OCR — zero extra deps on Windows 10+, via PowerShell.
4. Graceful failure with clear install hints.

Usage::

    from opendesk.computer.ocr import extract_text_from_region

    # region = (x, y, width, height) in logical pixels; None = full screen
    text = extract_text_from_region(region=(100, 200, 800, 400))
"""

from __future__ import annotations

import io
import os
import platform
import shutil
import subprocess
import tempfile
from pathlib import Path

_PLATFORM = platform.system()


def ocr_image(png_bytes: bytes, *, width: int = 0, height: int = 0) -> str:
    """Run OCR on a PNG byte buffer using the best available backend.

    Tries pytesseract, then platform-native OCR (Vision on macOS, WinRT on
    Windows).  Returns the extracted text or an informative error/install-hint
    string starting with ``"OCR "`` so callers can distinguish.

    ``width`` / ``height`` are optional; when set and small, the image is
    upscaled for better OCR accuracy.
    """
    try:
        import pytesseract  # type: ignore[import-not-found]
        from PIL import Image  # type: ignore[import-not-found]
        img = Image.open(io.BytesIO(png_bytes))
        w = width or img.width
        h = height or img.height
        if 0 < w < 300:
            factor = max(2, 300 // w)
            img = img.resize((w * factor, h * factor), Image.LANCZOS)
        text = pytesseract.image_to_string(img, config="--psm 6")
        return text.strip() or "(no text detected)"
    except ImportError:
        pass
    except Exception as exc:
        return f"pytesseract error: {exc}"

    if _PLATFORM == "Darwin":
        try:
            return _macos_vision_ocr(png_bytes)
        except Exception:
            pass

    if _PLATFORM == "Windows":
        try:
            return _windows_winrt_ocr(png_bytes)
        except Exception:
            pass

    return (
        "OCR not available. Install pytesseract:\n"
        "  macOS:   brew install tesseract && pip install pytesseract\n"
        "  Ubuntu:  sudo apt install tesseract-ocr && pip install pytesseract\n"
        "  Windows: choco install tesseract && pip install pytesseract\n\n"
        "(On macOS 11+ and Windows 10+ a built-in OCR engine is also tried "
        "automatically without any extra installs.)"
    )


def available_backend() -> str | None:
    """Return the name of the OCR backend :func:`ocr_image` would use, or
    ``None`` when none is usable.  Cheap: no image is processed."""
    try:
        import pytesseract  # type: ignore[import-not-found]
        from PIL import Image  # type: ignore[import-not-found]  # noqa: F401
        try:
            pytesseract.get_tesseract_version()
            return "pytesseract"
        except Exception:
            pass
    except ImportError:
        pass
    if _PLATFORM == "Darwin":
        if shutil.which("swiftc") or shutil.which("swift"):
            return "macos-vision"
    if _PLATFORM == "Windows":
        if shutil.which("powershell"):
            return "windows-winrt"
    return None


def warm_up() -> None:
    """Prepare the native backend ahead of time (compiles the macOS Vision
    helper on first use, which can take a minute).  No-op elsewhere."""
    if _PLATFORM == "Darwin":
        _compiled_vision_helper()


def extract_text_from_region(
    region: tuple[int, int, int, int] | None = None,
) -> str:
    """Capture a screen region and run OCR on it.

    Convenience wrapper that uses :func:`opendesk.computer.capture.capture_screen`
    plus :func:`ocr_image`.  Prefer calling them separately when integrating
    with a :class:`~opendesk.computer.Computer` so OCR can run on a captured
    :class:`~opendesk.computer.Pixmap` from any backend.
    """
    try:
        from opendesk.computer.capture import capture_screen
        png_bytes, w, h = capture_screen(region)
    except Exception as exc:
        return f"OCR error: could not capture screen: {exc}"
    return ocr_image(png_bytes, width=w, height=h)


_VISION_SWIFT_SRC = """
import Vision
import AppKit

let args = CommandLine.arguments
guard args.count > 1 else { exit(2) }
let url = URL(fileURLWithPath: args[1])
guard let img = NSImage(contentsOf: url),
      let cgImg = img.cgImage(forProposedRect: nil, context: nil, hints: nil)
else { exit(0) }

let req = VNRecognizeTextRequest()
req.recognitionLevel = .accurate
req.usesLanguageCorrection = true
let handler = VNImageRequestHandler(cgImage: cgImg, options: [:])
try? handler.perform([req])
let lines = (req.results ?? []).compactMap { $0.topCandidates(1).first?.string }
print(lines.joined(separator: "\\n"))
"""

_VISION_HELPER_VERSION = "1"


def _vision_helper_dir() -> Path:
    override = os.environ.get("OPENDESK_HOME")
    base = Path(override).expanduser() if override else Path.home() / ".opendesk"
    d = base / "bin"
    d.mkdir(parents=True, exist_ok=True)
    return d


def _compiled_vision_helper() -> Path | None:
    """Compile the Vision OCR helper once with ``swiftc`` and cache the
    binary under ``~/.opendesk/bin``.  Returns ``None`` if compilation
    isn't possible (no ``swiftc``), in which case callers fall back to
    interpreting the script with ``swift`` on every call."""
    swiftc = shutil.which("swiftc")
    if not swiftc:
        return None
    d = _vision_helper_dir()
    binary = d / f"vision-ocr-v{_VISION_HELPER_VERSION}"
    if binary.exists() and os.access(binary, os.X_OK):
        return binary
    src = d / "vision-ocr.swift"
    try:
        src.write_text(_VISION_SWIFT_SRC)
        r = subprocess.run(
            [swiftc, "-O", "-o", str(binary), str(src)],
            capture_output=True, text=True, timeout=180,
        )
        if r.returncode != 0 or not binary.exists():
            return None
        os.chmod(binary, 0o700)
        return binary
    except Exception:
        return None
    finally:
        src.unlink(missing_ok=True)


def _macos_vision_ocr(png_bytes: bytes) -> str:
    with tempfile.NamedTemporaryFile(suffix=".png", delete=False) as f:
        f.write(png_bytes)
        tmp_png = Path(f.name)

    try:
        helper = _compiled_vision_helper()
        if helper is not None:
            r = subprocess.run(
                [str(helper), str(tmp_png)], capture_output=True, text=True, timeout=30,
            )
            if r.returncode == 0:
                return r.stdout.strip() or "(no text detected)"
            # Fall through to the interpreted path on unexpected failure.

        swift_src = _VISION_SWIFT_SRC
        with tempfile.NamedTemporaryFile(suffix=".swift", delete=False, mode="w") as sf:
            sf.write(swift_src)
            swift_path = Path(sf.name)
        try:
            r = subprocess.run(
                ["swift", str(swift_path), str(tmp_png)],
                capture_output=True, text=True, timeout=60,
            )
            if r.returncode == 0:
                return r.stdout.strip() or "(no text detected)"
            raise RuntimeError(f"Swift Vision OCR failed: {r.stderr.strip()}")
        finally:
            swift_path.unlink(missing_ok=True)
    finally:
        tmp_png.unlink(missing_ok=True)


def _windows_winrt_ocr(png_bytes: bytes) -> str:
    with tempfile.NamedTemporaryFile(suffix=".png", delete=False) as f:
        f.write(png_bytes)
        tmp_png = Path(f.name)

    tmp_ps = str(tmp_png).replace("\\", "/")

    ps_script = f"""
Add-Type -AssemblyName System.Runtime.WindowsRuntime
[Windows.Storage.StorageFile, Windows.Storage, ContentType=WindowsRuntime] | Out-Null
[Windows.Media.Ocr.OcrEngine, Windows.Foundation, ContentType=WindowsRuntime] | Out-Null
[Windows.Graphics.Imaging.BitmapDecoder, Windows.Graphics, ContentType=WindowsRuntime] | Out-Null

$filePath = '{tmp_ps}'
$file = [Windows.Storage.StorageFile]::GetFileFromPathAsync($filePath).AsTask().Result
$stream = $file.OpenAsync([Windows.Storage.FileAccessMode]::Read).AsTask().Result
$decoder = [Windows.Graphics.Imaging.BitmapDecoder]::CreateAsync($stream).AsTask().Result
$bitmap = $decoder.GetSoftwareBitmapAsync().AsTask().Result
$engine = [Windows.Media.Ocr.OcrEngine]::TryCreateFromUserProfileLanguages()
$result = $engine.RecognizeAsync($bitmap).AsTask().Result
$result.Lines | ForEach-Object {{ $_.Text }}
"""
    try:
        r = subprocess.run(
            ["powershell", "-NonInteractive", "-Command", ps_script],
            capture_output=True, text=True, timeout=25,
        )
        if r.returncode == 0 and r.stdout.strip():
            return r.stdout.strip()
        raise RuntimeError(f"WinRT OCR failed: {r.stderr.strip()}")
    finally:
        tmp_png.unlink(missing_ok=True)
